### PR TITLE
Supporting OAuth 2.1 for public clients

### LINKED ISSUE #25316: Supporting OAuth 2.1 for public clients
### Description

Creating a default profile of client policy for Supporting OAuth 2.1 for public clients.

- The OAuth 2.1 Authorization Framework
https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-09

Its epic issue is https://github.com/keycloak/keycloak/issues/25141 .

### Discussion

_No response_

### Motivation

_No response_

### Details

It can be created by combining [OAuth 2.1 for confidential client](https://github.com/keycloak/keycloak/issues/25314) and [Sender-constrained tokens by DPoP](https://github.com/keycloak/keycloak/issues/25315) .

### PR DESCRIPTION
**Pull request by tnorimat**
closes #25316

### CODE REVIEW

**Comment by tnorimat** [comment]
@mposolda I added tests and documentation. Could you check the PR?

**Comment by tnorimat** [comment]
@mposolda I forgotten that DPoP is still a preview feature so that we cannot include its executor into a default profile until DPoP get out of a preview feature.
One option is that we would wait for completing https://github.com/keycloak/keycloak/issues/22311 . After that, this PR would be merged.
The other option is that this PR would exclude DPoP enforcer executor and a follow-up PR includes it after https://github.com/keycloak/keycloak/issues/22311 .

Do you think which would be a good option?

**Comment by mposolda** [comment]
> @mposolda I forgotten that DPoP is still a preview feature so that we cannot include its executor into a default profile until DPoP get out of a preview feature. One option is that we would wait for completing #22311 . After that, this PR would be merged. The other option is that this PR would exclude DPoP enforcer executor and a follow-up PR includes it after #22311 .
> 
> Do you think which would be a good option?

I am not sure when DPoP feature is promoted from preview...

Another option might be to use configuration including DPoP executor, but add something into the DPoP executor itself (instead of implementing `EnvironmentDependentFactory`). Perhaps something like:
```
If (Profile.isFeatureEnabled(DPOP) {
   executeChecksImpl(); // execute checks as done today
} else {
   logger.warnf("DPoP executor is used, but DPOP feature is not enabled. So DPOP is not enforced for the clients. Please enable DPOP feature in order to be able to have DPOP checks applied");
}
```

This way the configuration will work OOTB, but dpop checks just won't be done in runtime and instead warning would be shown if someone want to use oauth 2.1 client profile for public clients. WDYT?

**Review by andymunro — APPROVED** [review]

**Comment by tnorimat** [comment]
@mposolda I modified `DPoPBindEnforcerExecutor` and `DPoPBindEnforcerExecutorFactory` to call `Profile.isFeatureEnabled(Feature.DPOP)` in `DPoPBindEnforcerExecutor` and remove `EnvironmentDependentFactory` from  `DPoPBindEnforcerExecutorFactory`.

```
    @Override
    public void executeOnEvent(ClientPolicyContext context) throws ClientPolicyException {
        if (!Profile.isFeatureEnabled(Feature.DPOP)) {
            logger.warnf("DPoP executor is used, but DPOP feature is disabled. So DPOP is not enforced for the clients. " +
                    "Please enable DPOP feature in order to be able to have DPOP checks applied.");
            return;
        }

        HttpRequest request = session.getContext().getHttpRequest();
        switch (context.getEvent()) {
```

**Comment by mposolda** [comment]
@andymunro Thanks for the review!
@tnorimat Could you please rebase on top of latest main? I think that it will remove some (or all?) of documentation changes in this PR.

File: docs/documentation/securing_apps/topics/oidc/oauth21-support.adoc [hunk 1]
```diff
@@ -12,6 +12,6 @@ side may need to be still done manually or through some other third-party soluti
 ==== OAuth 2.1 client profiles
 
 To make sure that your clients are OAuth 2.1 compliant, you can configure Client Policies in your realm as described in the link:{adminguide_link}#_client_policies[{adminguide_name}]
-and link them to the global client profiles for OAuth 2.1 support, which are automatically available in each realm. You can use `oauth-2-1-for-confidential-client` profile for confidential clients.
+and link them to the global client profiles for OAuth 2.1 support, which are automatically available in each realm. You can use ether `oauth-2-1-for-confidential-client` profile for confidential clients or `oauth-2-1-for-public-client` profile for public clients.
```
Comment (mposolda) [inline docs/documentation/securing_apps/topics/oidc/oauth21-support.adoc:15]:
```suggestion
and link them to the global client profiles for OAuth 2.1 support, which are automatically available in each realm. You can use either `oauth-2-1-for-confidential-client` profile for confidential clients or `oauth-2-1-for-public-client` profile for public clients.
```

Small typo. Do you please have a chance to fix it?

**Comment by mposolda** [comment]
@tnorimat Thanks for the updates! Is it also possible to add some small note to the docs, that for OAuth 2.1 with public clients, it is recommended to use DPOP preview feature?

File: docs/documentation/securing_apps/topics/oidc/oauth21-support.adoc (hunk 1 above)
Comment (tnorimat) [reply docs/documentation/securing_apps/topics/oidc/oauth21-support.adoc:15]: Thank you. I fixed it.

**Comment by tnorimat** [comment]
@mposolda  I added the sentence to recommend using DPoP in oauth21-support.adoc.

> NOTE: When using OAuth 2.1 profile for public clients, it is recommended to use DPoP preview feature as described in the link:{adminguide_link}#_dpop-bound-tokens[{adminguide_name}] because DPoP binds an access token and a refresh token together with the public part of a client's key pair. This binding prevents an attacker from using stolen tokens.

**Review by mposolda — APPROVED** [review]
@tnorimat Thanks!

**Pull request merged**